socketserver/server/publisher.go

```diff
@@ -199,26 +240,50 @@ func HTTPBackendUncachedPublish(w http.ResponseWriter, r *http.Request) {
 
 	if cmd == "" {
 		w.WriteHeader(422)
-		fmt.Fprintf(w, "Error: cmd cannot be blank")
+		fmt.Fprint(w, "Error: cmd cannot be blank")
 		return
 	}
 	if channel == "" && scope != "global" {
 		w.WriteHeader(422)
-		fmt.Fprintf(w, "Error: channel must be specified")
+		fmt.Fprint(w, "Error: channel must be specified")
+		return
+	}
+	rl, err := rateLimitFromRequest(r)
+	if err != nil {
+		w.WriteHeader(422)
+		fmt.Fprintf(w, "error parsing ratelimit: %v", err)
 		return
 	}
 
 	cm := ClientMessage{MessageID: -1, Command: CommandPool.InternCommand(cmd), origArguments: json}
 	cm.parseOrigArguments()
-	var count int
 
-	switch scope {
-	default:
-		count = PublishToMultiple(strings.Split(channel, ","), cm)
-	case "global":
-		count = PublishToAll(cm)
+	var count int
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go rl.Run()
+	go func() {
+		switch scope {
+		default:
+			count = PublishToMultiple(strings.Split(channel, ","), cm, rl)
+		case "global":
+			count = PublishToAll(cm, rl)
+		}
+		wg.Done()
+		rl.Close()
+	}()
+	ch := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(ch)
+	}()
+	select {
+	case time.After(3 * time.Second):
+		count = -1
+	case <-ch:
 	}
```

Comment:
I can't help but feel there's a better way to do this. Seems overly complicated.